src/cryptonote_core/loki_name_system.cpp

```diff
@@ -0,0 +1,1124 @@
+      if (require_exact_len) err_stream << ", does not equal the required length=";
+      else                   err_stream <<" is 0 or exceeds the maximum length=";
+      err_stream << max << ", given value=";
+      err_stream.write(value, len);
+      *reason = err_stream.str();
+    }
+  }
+
+  return result;
+}
+
+bool validate_lns_value(cryptonote::network_type nettype, uint16_t type, char const *value, int value_len, lns_value *blob, std::string *reason)
+{
+  if (blob) *blob = {};
+  std::stringstream err_stream;
+
+  cryptonote::address_parse_info addr_info = {};
+
+  static_assert(GENERIC_VALUE_MAX >= MESSENGER_PUBLIC_KEY_BINARY_LENGTH, "lns_value assumes the largest blob size required, all other values should be able to fit into this buffer");
+  static_assert(GENERIC_VALUE_MAX >= LOKINET_ADDRESS_BINARY_LENGTH,      "lns_value assumes the largest blob size required, all other values should be able to fit into this buffer");
+  static_assert(GENERIC_VALUE_MAX >= sizeof(addr_info.address),          "lns_value assumes the largest blob size required, all other values should be able to fit into this buffer");
+  if (type == static_cast<uint16_t>(mapping_type::blockchain))
+  {
+    if (value_len == 0 || !get_account_address_from_str(addr_info, nettype, std::string(value, value_len)))
+    {
+      if (reason)
+      {
+        if (value_len == 0)
+        {
+          err_stream << "The value=";
+          err_stream.write(value, value_len);
+          err_stream << ", mapping into the wallet address, specifies a wallet address of 0 length";
+        }
+        else
+        {
+          err_stream << "Could not convert the wallet address string, check it is correct, value=";
+          err_stream.write(value, value_len);
+        }
+        *reason = err_stream.str();
+      }
+      return false;
+    }
+  }
+  else
+  {
+    int max_value_len            = lns::GENERIC_VALUE_MAX;
+    bool value_require_exact_len = true;
+    if (type == static_cast<uint16_t>(mapping_type::lokinet))        max_value_len = (LOKINET_ADDRESS_BINARY_LENGTH * 2);
+    else if (type == static_cast<uint16_t>(mapping_type::messenger)) max_value_len = (MESSENGER_PUBLIC_KEY_BINARY_LENGTH * 2);
+    else value_require_exact_len = false;
+
+    if (!check_lengths(type, value, value_len, max_value_len, value_require_exact_len, reason))
+      return false;
+  }
+
+  if (type == static_cast<uint16_t>(mapping_type::blockchain))
+  {
+    if (blob)
+    {
+      blob->len = sizeof(addr_info.address);
+      memcpy(blob->buffer.data(), &addr_info.address, blob->len);
+    }
+  }
+  else
+  {
+    // NOTE: Check value is hex
+    if ((value_len % 2) != 0)
+    {
+      if (reason)
+      {
+        err_stream << "The value=";
+        err_stream.write(value, value_len);
+        err_stream << ", should be a hex string that has an even length to be convertible back into binary, length=" << value_len;
+        *reason = err_stream.str();
```

Comment:
lokinet addresses are never represented anywhere in hex, so it's going to be a little difficult to get users to provide it this way.  Instead we should require that for lokinet the value is the z-base32-encoded value that lokinet uses: it's still a 32-byte value, but is represented 52 characters long (not including the ".loki") and has values in `ybndrfg8ejkmcpqxot1uwisza345h769`.  You can lift [the encode/decode functions out of lokinet](https://github.com/loki-project/loki-network/blob/master/llarp/util/encode.hpp).